2.8 PHP7 compatability

dereuromark opened the pull request:
Resolves https://github.com/cakephp/cakephp/issues/7671 with BC.

I opened a PR to be able to work together making CakePHP 2.8+ PHP7 ready.
Anyone who can contribute, please feel free to add commits on top.
Right now it segfaults or dies away at 27%. Ideas are welcome.

Commit:
Enable PHP7.0 in travis

Commit:
Allow ParseError as exception and remove typehint.

Commit:
Remove another type hint

Commit:
Skip error for now.

kfrijters committed:
Removed last typehint Exception

#7671 Removed last typehint Exception as it was still giving an uncaught typeError message in PHP7

dereuromark committed:
Merge pull request #7807 from m1nd53t/2.8-PHP7

Removed last typehint Exception

markstory commented:
Have you run the suite with `--debug` yet? That will output the test method names as methods are run.

phpnut commented:
I think I found the issues stopping test for continuing past the 27% @dereuromark mentioned. There are a lot of places in the tests where null is being use when a class or interface is expected. Will continue later today and push changes

Commit:
Allowing tests to run on PHP 7

Commit:
FIxing srand() expects parameter 1 to be integer, string given

Type casting to integer

Commit:
add @throws anotation to fix travis PHP_CODESNIFFER warnings

phpnut committed:
Adding missing @return tag in function comment

phpnut committed:
Forcing bool return

phpnut committed:
Forcing bool

phpnut committed:
Changing where bool is returned

phpnut committed:
Using better way to check for active session

phpnut committed:
Correcting return

phpnut committed:
session_status() not available until 5.4

Adding check for php version

phpnut committed:
Revert "Adding missing @return tag in function comment"

This reverts commit 130e854c30f11a125dbc4ce783e370a7df6b81fb.

phpnut committed:
Revert "Revert "Adding missing @return tag in function comment""

This reverts commit 1cc5c8cf7fe131f1389391fcd83f02168ba07a74.

phpnut committed:
Revert "session_status() not available until 5.4"

This reverts commit da9b6453cee2f025aecb9ab30c9d3029fbf527db.

phpnut committed:
Revert "Correcting return"

This reverts commit 6e09e64432e5d517c74970d5ae9bbe49ab2db192.

phpnut committed:
Revert "Using better way to check for active session"

This reverts commit 439a33c70827d0986986ca9ad8c3b3daf94fb84f.

phpnut committed:
Revert "Changing where bool is returned"

This reverts commit 8f52cefdccceb44a5e2e705261780a95d971c29c.

phpnut committed:
Revert "Forcing bool"

This reverts commit 26ea74cbcff8690f04686940dc7dff708744642d.

phpnut committed:
Revert "Forcing bool return"

This reverts commit fac95baee7de3e40c241cca8ab8386607d547411.

phpnut committed:
Adding boolean return in Cache::gc

phpnut committed:
Changing return types

read now returns empty string instead of false when read method returns an empty value.
write, destroy and gc will return boolean type

phpnut committed:
Adding return types to Session test classes.

phpnut committed:
Removing forced return true

sebastienbarre commented:
Look at the latest [Travis CI build](https://travis-ci.org/cakephp/cakephp/builds/96883308), it seems to pass the tests, or am I missing something? Just curious.

dereuromark commented:
@sebastienbarre No, they are failing as you can see when you look into the PHP7 one, but for some reason travis marks them as passed.

markstory commented:
[Session tests are pretty sad](https://travis-ci.org/cakephp/cakephp/jobs/96883302) I think @phpnut was close to having a solution on these though.

phpnut committed:
Should be able to write a file with no data in it.

phpnut committed:
Fixes session_write_close(): Skipping numeric key 0 error

phpnut committed:
Fixes tests to expect changes made to read and write methods - This could be a possible BC change

Since php 7 expects write to return true or false this needed to change, previous implementation would return the values sent to write on success and false on failure. Similar change to read method test CakeSession::read() now returns results or ''.

phpnut committed:
Initial fixes for failing session tests

phpnut committed:
Refactoring and emoving one isset check

phpnut committed:
Fixing Code Sniffer errors

phpnut committed:
Reverted change setting $_SESSION to an array. Commenting out a test that is invalid.

This test creates a numeric key of 0 in $_SESSION which is not a valid session key. This causes error - session_write_close(): Skipping numeric key 0 error.

phpnut committed:
Fixes SessionComponentTest::testSessionValid. Refactored _hasSession and other erros on php 7

init would always set CakeSession::$_userAgent.

phpnut committed:
Fixes: A non well formed numeric value encountered - php 7

phpnut committed:
Fixing code sniffer errors

phpnut commented:
If this looks ok I will merge to 2.8

markstory commented on lib/Cake/Cache/Engine/FileEngine.php:
```diff
@@ -109,7 +109,7 @@ public function gc($expires = null) {
  * @return bool True if the data was successfully cached, false on failure
  */
 	public function write($key, $data, $duration) {
-		if ($data === '' || !$this->_init) {
```
This will result in cache files being written for '' content. But I think that is OK and having PHP7 compatibility is more important. We might want to document this in the 2.8 migration guide though.

markstory commented on lib/Cake/Model/Datasource/CakeSession.php:
```diff
@@ -218,6 +217,9 @@ public static function start() {
  */
 	public static function started() {
+		if (PHP_VERSION >= 5.4) {
+			return isset($_SESSION) && (session_status() === PHP_SESSION_ACTIVE);
```
Instead of sniffing the version, could we check for the existence of `session_status()`?

markstory commented on lib/Cake/Model/Datasource/CakeSession.php:
```diff
@@ -461,7 +463,14 @@ public static function destroy() {
+			}
 			session_destroy();
+			if (isset($_COOKIE[static::_cookieName()])) {
+				unset($_COOKIE[static::_cookieName()]);
```
`unset()` should be safe to call on keys that are not defined.

markstory commented on lib/Cake/Model/Datasource/Session/CacheSession.php:
```diff
@@ -52,7 +52,11 @@ public function close() {
 	public function read($id) {
-		return Cache::read($id, Configure::read('Session.handler.config'));
+		$data = Cache::read($id, Configure::read('Session.handler.config'));
+		if (empty($data)) {
```
Can the data ever be '0'?

markstory commented on lib/Cake/Test/Case/Controller/Component/SessionComponentTest.php:
```diff
@@ -202,7 +202,8 @@ public function testSessionReadWrite() {
 
-		$this->assertTrue($Session->write(array('Test'), 'some value'));
+		//This test creates a numeric key 0 which is not a valid session key
+		//$this->assertTrue($Session->write(array('Test'), 'some value'));
```
I would just remove this.

phpnut committed:
Allowing returning numeric 0 from read

phpnut committed:
Removing check for set key

phpnut committed:
Fixing Undefined index: Session

phpnut committed:
Removing invalid test

phpnut committed:
Using function_exists instead of using constant

phpnut committed:
getting sloppy as I get older, fixing code sniffer errors
